README.md
# D3toD4

This project was generated with [Angular CLI](https://github.com/angular/angular-cli) version 1.1.3.

## Development server

Run `ng serve` for a dev server. Navigate to `http://localhost:4200/`. The app will automatically reload if you change any of the source files.

## Code scaffolding

Run `ng generate component component-name` to generate a new component. You can also use `ng generate directive|pipe|service|class|module`.

## Build

Run `ng build` to build the project. The build artifacts will be stored in the `dist/` directory. Use the `-prod` flag for a production build.

## Running unit tests

Run `ng test` to execute the unit tests via [Karma](https://karma-runner.github.io).

## Running end-to-end tests

Run `ng e2e` to execute the end-to-end tests via [Protractor](http://www.protractortest.org/).
Before running the tests make sure you are serving the app via `ng serve`.

## Further help

To get more help on the Angular CLI use `ng help` or go check out the [Angular CLI README](https://github.com/angular/angular-cli/blob/master/README.md).
"# POCChartD3" 


docs/changelog.md
# Changelog

## 6.1.0
- Feature: Added inputs to control minimum and maximum values on the axes (#582, #359)
- Feature: Added input to enable/disable animations (#565)
- Feature: Added designatedTotal Input for Pie Grid (#554)
- Enhancement: Don't transform text to uppercase in legends and labels (#578)
- Enhancement: Changed legends and label colors (#578)
- Enhancement: Increased font size in advanced legend and number cards subtext (#578)
- Enhancement: Hide legend title when title is missing (#578)
- Enhancement: Added x-axis value to the context of a bubble chart tooltip (#530)
- Bug: Removed path from SVG refs to fix gradients not working when URL has queryString (#584) 
- Bug: Fixed bug where value arc animation on gauge chart would start from 0 on update (#563)
- Bug: Removed date conversion for name-value at series-tooltip (#544)
- Bug: Added null check for rootComponents in injection servoce (#574)
- Bug: Fixed issue where the tooltip would not appear for certain points (#501)
- Bug: Fixed issue where pie chart colors would not update on color changes (#551)
- Bug: Added polyfill for SVGElement.prototype.contains (#386)
- Bug: Renamed ngOutletContext (deprecated) to ngTemplateOutletContext (#535)
- Docs: Added custom chart page

## 6.0.1
- Bug: Fixed line chart stroke color
- Bug: Fixed timeline width and alignment
- Bug: Fixed truncation of legend text (#487)
- Chore: Removed patch numbers from peerDependencies (#478)

## 6.0.0
- Breaking: Removed `showSeriesOnHover` option of line-chart
- Refactor: Improved performance of line and area chart tooltips (#463)
- Enhancement: Added `roundEdges` input property to bar charts (#408)
- Enhancement: Tooltip elements are not created if tooltips are disabled (#415)
- Enhancement: Added the option to define reference lines in line charts (#422)
- Enhancement: Improved animations
- Bug: Fixed timeline select issue (#425)
- Chore: Upgraded angular to 4.2.5
- Demo: Added timeline filter bar chart demo
- Demo: Added Combo chart (bar and line) demo (#432)

## 5.3.1
- Bug: Fix bug where the axis component would not emit a dimensionsChanged event in some cases

## 5.3.0
- Feature: Added option to pass custom ng-templates for tooltips to all charts
- Enhancement: Added 'label' input to advanced pie chart component
- Enhancement: Exposed xOrient and yOrient properties on axis components
- Enhancement: Set default dimensions if width or height are undefined or 0
- Enhancement: Set a default color scheme
- Bug: Fixed bubble chart overriding the view input
- Bug: Fixed linear color scheme when there are less than 3 colors
- Bug: Fixed gradient not applying to straight lines
- Demo: Added sparkline custom demo chart
- Demo: Added interactive treemap demo chart

## 5.2.1
- Bug: Remove RGBColor type from color-util file (#394)

## 5.2.0
- Feature: Added Polar/Radar chart type (#380)
- Enhancement: Disable pointer if no click events are attached to number cards (#378)
- Enhancement: Added gradiant to tree map chart (#382)
- Enhancement: Added support for inverted colors from rgba values in number cards (#366)
- Enhancement: Added value and label formatting to tree map (#383)
- Refactor: loremess restrictive angular dependency versioning (#358)
- Bug: Fixed types in gauge chart (#364)
- Bug: Fixed axis and tooltips on bubble chart (#352)
- Bug: Added styleUrls and changeDetection to bubble chart
- Bug: Truncate dimension values
- Bug: Fixed number card count animation (#387)
- Bug: Added type to gauge chart valueFormatting parameter (#388)

## 5.1.2
- Bug: Fix number card scaling in hybrid apps using ngUpgrade  (#338)
- Bug: Fix label trimming and clipped SVGs (#340)

## 5.1.1
- Bug: Add default value for label format  (#334)

## 5.1.0
- Enhancement: Added valueFormatting input to number card chart (#325)
- Enhancement: Added label formatting input to number card chart (#332)
- Enhancement: Added tooltipText to pie charts (#331)
- Enhancement: Added option in line-chart for disabling series on hover (#323)
- Enhancement: Added an input for custom legend title (#317)
- Bug: Fixed number card text scaling when values change (#330)
- Bug: Inverted y-axis on horizontal bar charts (#328)
- Bug: Fix division by zero in pie label component

## 5.0.1
- Enhancement: `valueFormatting` applies to Gauge Arc Labels too

## 5.0.0
- Breaking: Upgrade to angular 4

## 4.4.0
- Feature: Added ability to set card text color in number card (#284)
- Enhancement: Added more heatmap options and support for various inputs for innerPadding (#297, #301)
- Enhancement: Added valueFormatting input to linear gauge chart
- Enhancement: Added valueFormatting input to gauge chart
- Enhancement: Added labelFormatting input to pie chart
- Enhancement: Updated pie chart label animations (#291)
- Bug: Fixed pie chart positioning
- Bug: Fixed ticks calculation on resize (#295)
- Bug: Fixed percentage value in pie grid when the total is 0 (#288)
- Bug: Fixed date formatting in bubble chart tooltips (#285)
- Bug: Fixed Scale.round error in line chart ()#287)
- Chore: AOT error fixes

## 4.3.0
- Enhancement: Improved pie chart label positioning (#273)
- Enhancement: Improved number card design and scaling (#272)
- Chore: Refactored charts to use d3 modules directly in library components (#270)

## 4.2.1
- Bug: Fixed customColors mapping
- Chore: Various AoT compilation and build improvements (#255, #257, #258, #259, #256, #263, #260, #266)

## 4.2.0
- Chore: AoT compilation compatibility improvements (#253, #252, #251)
- Enhancement: Added option to disable tooltips (#223)

## 4.1.3
- Enhancement: Added option to override the value in gauge chart
- Bug: Fixed padding on bubble charts
- Chore: Compile SCSS into CSS before building
- Chore: Expor ChartComponent through the main module

## 4.1.2
- Bug: Fix tooltip when axis labels are not shown in bubble chart
- Bug: Do not round scales when they are not linear

demo/data/data.ts
// C-SK
// import { data as countries } from 'emoji-flags';

export const single = [
  {
    name: 'Germany',
    value: 40632
  },
  {
    name: 'United States',
    value: 49737
  },
  {
    name: 'France',
    value: 36745
  },
  {
    name: 'United Kingdom',
    value: 36240
  },
  {
    name: 'Spain',
    value: 33000
  },
  {
    name: 'Italy',
    value: 35800
  }
];

export const multi = [
  {
    name: 'Germany',
    series: [
      {
        name: '2010',
        value: 40632,
        series: single
      },
      {
        name: '2000',
        value: 36953,
        series: single
      },
      {
        name: '1990',
        value: 31476,
        series: single
      }
    ]
  },
  {
    name: 'United States',
    series: [
      {
        name: '2010',
        value: 49737,
        series: single
      },
      {
        name: '2000',
        value: 45986,
        series: single
      },
      {
        name: '1990',
        value: 37060,
        series: single
      }
    ]
  },
  {
    name: 'France',
    series: [
      {
        name: '2010',
        value: 36745,
        series: single
      },
      {
        name: '2000',
        value: 34774,
        series: single
      },
      {
        name: '1990',
        value: 29476,
        series: single
      }
    ]
  },
  {
    name: 'United Kingdom',
    series: [
      {
        name: '2010',
        value: 36240,
        series: single
      },
      {
        name: '2000',
        value: 32543,
        series: single
      },
      {
        name: '1990',
        value: 26424,
        series: single
      }
    ]
  }
];

export const bubble = [
  {
    name: 'Germany',
    series: [
      {
        name: '2010',
        x: new Date(2010, 0, 1),
        y: 800.3,
        r: 80.4
      },
      {
        name: '2000',
        x: new Date(2000, 0, 1),
        y: 8.3,
        r: 78
      },
      {
        name: '1990',
        x: new Date(1990, 0, 1),
        y: 75.4,
        r: 79
      }
    ]
  },
  {
    name: 'United States',
    series: [
      {
        name: '2010',
        x: new Date(2010, 0, 1),
        y: 78.8,
        r: 310
      },
      {
        name: '2000',
        x: new Date(2000, 0, 1),
        y: 7.9,
        r: 283
      },
      {
        name: '1990',
        x: new Date(1990, 0, 1),
        y: 500.4,
        r: 23
      }
    ]
  },
  {
    name: 'France',
    series: [
      {
        name: '2010',
        x: new Date(2010, 0, 1),
        y: 81.4,
        r: 63
      },
      {
        name: '2000',
        x: new Date(2000, 0, 1),
        y: 790.1,
        r: 59.4
      },
      {
        name: '1990',
        x: new Date(1990, 0, 1),
        y: 10.2,
        r: 56.9
      }
    ]
  },
  {
    name: 'United Kingdom',
    series: [
      {
        name: '2010',
        x: new Date(2010, 0, 1),
        y: 80.2,
        r: 62.7
      },
      {
        name: '2000',
        x: new Date(2000, 0, 1),
        y: 77.8,
        r: 58.9
      },
      {
        name: '1990',
        x: new Date(1990, 0, 1),
        y: 75.7,
        r: 57.1
      }
    ]
  }
];

export function generateGraph(nodeCount: number) {
  // C-SK
  const countries: any = []
  const nodes = [];
  const links = [];
  for (let i = 0; i < nodeCount; i++) {
    const country = countries[Math.floor(Math.random() * countries.length)];
    nodes.push({
      value: country.name,
    });
    for (let j = 0; j < nodes.length - 1; j++) {
      if (Math.random() < 0.03) {
        links.push({
          source: country,
          target: nodes[j].value,
        });
      }
    }
  }
  return { links, nodes };
}

export function timelineFilterBarData() {
  const results: any[] = [];
  const dataPoints = 30;
  const domain: Date[] = []; // array of time stamps in milliseconds
  const dayLength = 24 * 60 * 60 * 1000;
  let date = 1473700105009; // Sep 12, 2016
  for (let j = 0; j < dataPoints; j++) {
    // random dates between Sep 12, 2016 and Sep 24, 2016
    results.push({
      name: new Date(date),
      value: Math.floor(Math.random() * 300)
    });
    date += dayLength;
  }

  return results;
}

export function generateData(seriesLength: number, includeMinMaxRange: boolean, dataPoints: number = 5): any[] {
  const results = [];

  const domain: Date[] = []; // array of time stamps in milliseconds

  for (let j = 0; j < dataPoints; j++) {
    // random dates between Sep 12, 2016 and Sep 24, 2016
    domain.push(new Date(Math.floor(1473700105009 + Math.random() * 1000000000)));
  }
  // C-SK;
  const countries: any = [];
  for (let i = 0; i < seriesLength; i++) {
    const country = countries[Math.floor(Math.random() * countries.length)];
    const series = {
      name: country.name,
      series: []
    };

    for (let j = 0; j < domain.length; j++) {
      const value = Math.floor(2000 + Math.random() * 5000);
      // let timestamp = Math.floor(1473700105009 + Math.random() * 1000000000);
      const timestamp = domain[j];
      if (includeMinMaxRange) {
        const errorMargin = 0.02 + Math.random() * 0.08;

        series.series.push({
          value,
          name: timestamp,
          min: Math.floor(value * (1 - errorMargin)),
          max: Math.ceil(value * (1 + errorMargin))
        });
      } else {
        series.series.push({
          value,
          name: timestamp
        });
      }
    }

    results.push(series);
  }
  return results;
}

export const treemap = [{
  name: 'flare',
  children: [
    {
      name: 'analytics',
      children: [
        {
          name: 'cluster',
          children: [
            { name: 'AgglomerativeCluster', size: 3938 },
            { name: 'CommunityStructure', size: 3812 },
            { name: 'HierarchicalCluster', size: 6714 },
            { name: 'MergeEdge', size: 743 }
          ]
        },
        {
          name: 'graph',
          children: [
            { name: 'BetweennessCentrality', size: 3534 },
            { name: 'LinkDistance', size: 5731 },
            { name: 'MaxFlowMinCut', size: 7840 },
            { name: 'ShortestPaths', size: 5914 },
            { name: 'SpanningTree', size: 3416 }
          ]
        },
        {
          name: 'optimization',
          children: [
            { name: 'AspectRatioBanker', size: 7074 }
          ]
        }
      ]
    },
    {
      name: 'animate',
      children: [
        { name: 'Easing', size: 17010 },
        { name: 'FunctionSequence', size: 5842 },
        {
          name: 'interpolate',
          children: [
            { name: 'ArrayInterpolator', size: 1983 },
            { name: 'ColorInterpolator', size: 2047 },
            { name: 'DateInterpolator', size: 1375 },
            { name: 'Interpolator', size: 8746 },
            { name: 'MatrixInterpolator', size: 2202 },
            { name: 'NumberInterpolator', size: 1382 },
            { name: 'ObjectInterpolator', size: 1629 },
            { name: 'PointInterpolator', size: 1675 },
            { name: 'RectangleInterpolator', size: 2042 }
          ]
        },
        { name: 'ISchedulable', size: 1041 },
        { name: 'Parallel', size: 5176 },
        { name: 'Pause', size: 449 },
        { name: 'Scheduler', size: 5593 },
        { name: 'Sequence', size: 5534 },
        { name: 'Transition', size: 9201 },
        { name: 'Transitioner', size: 19975 },
        { name: 'TransitionEvent', size: 1116 },
        { name: 'Tween', size: 6006 }
      ]
    },
    {
      name: 'data',
      children: [
        {
          name: 'converters',
          children: [
            { name: 'Converters', size: 721 },
            { name: 'DelimitedTextConverter', size: 4294 },
            { name: 'GraphMLConverter', size: 9800 },
            { name: 'IDataConverter', size: 1314 },
            { name: 'JSONConverter', size: 2220 }
          ]
        },
        { name: 'DataField', size: 1759 },
        { name: 'DataSchema', size: 2165 },
        { name: 'DataSet', size: 586 },
        { name: 'DataSource', size: 3331 },
        { name: 'DataTable', size: 772 },
        { name: 'DataUtil', size: 3322 }
      ]
    },
    {
      name: 'display',
      children: [
        { name: 'DirtySprite', size: 8833 },
        { name: 'LineSprite', size: 1732 },
        { name: 'RectSprite', size: 3623 },
        { name: 'TextSprite', size: 10066 }
      ]
    },
    {
      name: 'flex',
      children: [
        { name: 'FlareVis', size: 4116 }
      ]
    },
    {
      name: 'physics',
      children: [
        { name: 'DragForce', size: 1082 },
        { name: 'GravityForce', size: 1336 },
        { name: 'IForce', size: 319 },
        { name: 'NBodyForce', size: 10498 },
        { name: 'Particle', size: 2822 },
        { name: 'Simulation', size: 9983 },
        { name: 'Spring', size: 2213 },
        { name: 'SpringForce', size: 1681 }
      ]
    },
    {
      name: 'query',
      children: [
        { name: 'AggregateExpression', size: 1616 },
        { name: 'And', size: 1027 },
        { name: 'Arithmetic', size: 3891 },
        { name: 'Average', size: 891 },
        { name: 'BinaryExpression', size: 2893 },
        { name: 'Comparison', size: 5103 },
        { name: 'CompositeExpression', size: 3677 },
        { name: 'Count', size: 781 },
        { name: 'DateUtil', size: 4141 },
        { name: 'Distinct', size: 933 },
        { name: 'Expression', size: 5130 },
        { name: 'ExpressionIterator', size: 3617 },
        { name: 'Fn', size: 3240 },
        { name: 'If', size: 2732 },
        { name: 'IsA', size: 2039 },
        { name: 'Literal', size: 1214 },
        { name: 'Match', size: 3748 },
        { name: 'Maximum', size: 843 },
        {
          name: 'methods',
          children: [
            { name: 'add', size: 593 },
            { name: 'and', size: 330 },
            { name: 'average', size: 287 },
            { name: 'count', size: 277 },
            { name: 'distinct', size: 292 },
            { name: 'div', size: 595 },
            { name: 'eq', size: 594 },
            { name: 'fn', size: 460 },
            { name: 'gt', size: 603 },
            { name: 'gte', size: 625 },
            { name: 'iff', size: 748 },
            { name: 'isa', size: 461 },
            { name: 'lt', size: 597 },
            { name: 'lte', size: 619 },
            { name: 'max', size: 283 },
            { name: 'min', size: 283 },
            { name: 'mod', size: 591 },
            { name: 'mul', size: 603 },
            { name: 'neq', size: 599 },
            { name: 'not', size: 386 },
            { name: 'or', size: 323 },
            { name: 'orderby', size: 307 },
            { name: 'range', size: 772 },
            { name: 'select', size: 296 },
            { name: 'stddev', size: 363 },
            { name: 'sub', size: 600 },
            { name: 'sum', size: 280 },
            { name: 'update', size: 307 },
            { name: 'variance', size: 335 },
            { name: 'where', size: 299 },
            { name: 'xor', size: 354 },
            { name: '_', size: 264 }
          ]
        },
        { name: 'Minimum', size: 843 },
        { name: 'Not', size: 1554 },
        { name: 'Or', size: 970 },
        { name: 'Query', size: 13896 },
        { name: 'Range', size: 1594 },
        { name: 'StringUtil', size: 4130 },
        { name: 'Sum', size: 791 },
        { name: 'Variable', size: 1124 },
        { name: 'Variance', size: 1876 },
        { name: 'Xor', size: 1101 }
      ]
    },
    {
      name: 'scale',
      children: [
        { name: 'IScaleMap', size: 2105 },
        { name: 'LinearScale', size: 1316 },
        { name: 'LogScale', size: 3151 },
        { name: 'OrdinalScale', size: 3770 },
        { name: 'QuantileScale', size: 2435 },
        { name: 'QuantitativeScale', size: 4839 },
        { name: 'RootScale', size: 1756 },
        { name: 'Scale', size: 4268 },
        { name: 'ScaleType', size: 1821 },
        { name: 'TimeScale', size: 5833 }
      ]
    },
    {
      name: 'util',
      children: [
        { name: 'Arrays', size: 8258 },
        { name: 'Colors', size: 10001 },
        { name: 'Dates', size: 8217 },
        { name: 'Displays', size: 12555 },
        { name: 'Filter', size: 2324 },
        { name: 'Geometry', size: 10993 },
        {
          name: 'heap',
          children: [
            { name: 'FibonacciHeap', size: 9354 },
            { name: 'HeapNode', size: 1233 }
          ]
        },
        { name: 'IEvaluable', size: 335 },
        { name: 'IPredicate', size: 383 },
        { name: 'IValueProxy', size: 874 },
        {
          name: 'math',
          children: [
            { name: 'DenseMatrix', size: 3165 },
            { name: 'IMatrix', size: 2815 },
            { name: 'SparseMatrix', size: 3366 }
          ]
        },
        { name: 'Maths', size: 17705 },
        { name: 'Orientation', size: 1486 },
        {
          name: 'palette',
          children: [
            { name: 'ColorPalette', size: 6367 },
            { name: 'Palette', size: 1229 },
            { name: 'ShapePalette', size: 2059 },
            { name: 'SizePalette', size: 2291 }
          ]
        },
        { name: 'Property', size: 5559 },
        { name: 'Shapes', size: 19118 },
        { name: 'Sort', size: 6887 },
        { name: 'Stats', size: 6557 },
        { name: 'Strings', size: 22026 }
      ]
    },
    {
      name: 'vis',
      children: [
        {
          name: 'axis',
          children: [
            { name: 'Axes', size: 1302 },
            { name: 'Axis', size: 24593 },
            { name: 'AxisGridLine', size: 652 },
            { name: 'AxisLabel', size: 636 },
            { name: 'CartesianAxes', size: 6703 }
          ]
        },
        {
          name: 'controls',
          children: [
            { name: 'AnchorControl', size: 2138 },
            { name: 'ClickControl', size: 3824 },
            { name: 'Control', size: 1353 },
            { name: 'ControlList', size: 4665 },
            { name: 'DragControl', size: 2649 },
            { name: 'ExpandControl', size: 2832 },
            { name: 'HoverControl', size: 4896 },
            { name: 'IControl', size: 763 },
            { name: 'PanZoomControl', size: 5222 },
            { name: 'SelectionControl', size: 7862 },
            { name: 'TooltipControl', size: 8435 }
          ]
        },
        {
          name: 'data',
          children: [
            { name: 'Data', size: 20544 },
            { name: 'DataList', size: 19788 },
            { name: 'DataSprite', size: 10349 },
            { name: 'EdgeSprite', size: 3301 },
            { name: 'NodeSprite', size: 19382 },
            {
              name: 'render',
              children: [
                { name: 'ArrowType', size: 698 },
                { name: 'EdgeRenderer', size: 5569 },
                { name: 'IRenderer', size: 353 },
                { name: 'ShapeRenderer', size: 2247 }
              ]
            },
            { name: 'ScaleBinding', size: 11275 },
            { name: 'Tree', size: 7147 },
            { name: 'TreeBuilder', size: 9930 }
          ]
        },
        {
          name: 'events',
          children: [
            { name: 'DataEvent', size: 2313 },
            { name: 'SelectionEvent', size: 1880 },
            { name: 'TooltipEvent', size: 1701 },
            { name: 'VisualizationEvent', size: 1117 }
          ]
        },
        {
          name: 'legend',
          children: [
            { name: 'Legend', size: 20859 },
            { name: 'LegendItem', size: 4614 },
            { name: 'LegendRange', size: 10530 }
          ]
        },
        {
          name: 'operator',
          children: [
            {
              name: 'distortion',
              children: [
                { name: 'BifocalDistortion', size: 4461 },
                { name: 'Distortion', size: 6314 },
                { name: 'FisheyeDistortion', size: 3444 }
              ]
            },
            {
              name: 'encoder',
              children: [
                { name: 'ColorEncoder', size: 3179 },
                { name: 'Encoder', size: 4060 },
                { name: 'PropertyEncoder', size: 4138 },
                { name: 'ShapeEncoder', size: 1690 },
                { name: 'SizeEncoder', size: 1830 }
              ]
            },
            {
              name: 'filter',
              children: [
                { name: 'FisheyeTreeFilter', size: 5219 },
                { name: 'GraphDistanceFilter', size: 3165 },
                { name: 'VisibilityFilter', size: 3509 }
              ]
            },
            { name: 'IOperator', size: 1286 },
            {
              name: 'label',
              children: [
                { name: 'Labeler', size: 9956 },
                { name: 'RadialLabeler', size: 3899 },
                { name: 'StackedAreaLabeler', size: 3202 }
              ]
            },
            {
              name: 'layout',
              children: [
                { name: 'AxisLayout', size: 6725 },
                { name: 'BundledEdgeRouter', size: 3727 },
                { name: 'CircleLayout', size: 9317 },
                { name: 'CirclePackingLayout', size: 12003 },
                { name: 'DendrogramLayout', size: 4853 },
                { name: 'ForceDirectedLayout', size: 8411 },
                { name: 'IcicleTreeLayout', size: 4864 },
                { name: 'IndentedTreeLayout', size: 3174 },
                { name: 'Layout', size: 7881 },
                { name: 'NodeLinkTreeLayout', size: 12870 },
                { name: 'PieLayout', size: 2728 },
                { name: 'RadialTreeLayout', size: 12348 },
                { name: 'RandomLayout', size: 870 },
                { name: 'StackedAreaLayout', size: 9121 },
                { name: 'TreeMapLayout', size: 9191 }
              ]
            },
            { name: 'Operator', size: 2490 },
            { name: 'OperatorList', size: 5248 },
            { name: 'OperatorSequence', size: 4190 },
            { name: 'OperatorSwitch', size: 2581 },
            { name: 'SortOperator', size: 2023 }
          ]
        },
        { name: 'Visualization', size: 16540 }
      ]
    }
  ]
}];
